### PR TITLE
Move message disposition methods to Receiver

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
Moved methods from Message to Receiver.
Removed Receiver.HandleMessage.
Fixed auto-accept in ModeFirst.
Check for closed link during blocking Receive.

### CODE REVIEW

File: receiver.go [hunk 1]
```diff
@@ -117,7 +61,7 @@ func (r *Receiver) Receive(ctx context.Context) (*Message, error) {
 		// and keeps the behavior the same as before the unsettled messages tracking was introduced
 		defer r.link.DeleteUnsettled(&msg)
```
Comment (serbrech) [inline receiver.go:62]: this breaks the modesecond. that's the whole point of the HandleMessage api. you can't delete the message as soon as you return it to the caller, it's not disposed yet, so it's not settled, and the credit should not be issued.

File: receiver.go [hunk 2]
```diff
@@ -26,64 +26,6 @@ type Receiver struct {
...
-// When using ModeFirst, the message is spontaneously Accepted at reception.
-func (r *Receiver) HandleMessage(ctx context.Context, handle func(*Message) error) error {
-	debug(3, "Entering link %s Receive()", r.link.Key.name)
-
-	trackCompletion := func(msg *Message) {
-		if msg.doneSignal == nil {
-			msg.doneSignal = make(chan struct{})
-		}
-		<-msg.doneSignal
-		r.link.DeleteUnsettled(msg)
```
Comment (serbrech) [inline receiver.go:42]: this is the reason for the handler.
you can only remove the message from the map tracking it on settlement, and not before.
otherwise you are receiving an unlimited messages concurrently, no matter the link credit tha tyou initially configure, because the map tracking the message being handled is cleared as soon as you return the message.

Comment (richardpark-msft) [reply receiver.go:42]: This is an interesting one. There are two credit management algorithms now (manual and the previous one). 

I think we _only_ need to activate this logic when we're **not** doing manual credit management. So I should have extended some additional changes (probably to just not track these 'unsettled' messages in the first place if a manualCreditor exists).

File: receiver.go (hunk 1 above)
Comment (richardpark-msft) [reply receiver.go:62]: Yeah, I think we need to do a manual creditor check here too and opt out if it's enabled. 

Ie, tracking unsettled messages is the "autocredit management" path.

Comment (jhendrixMSFT) [reply receiver.go:62]: It still needs to move though.  We can't remove the message from the unsettled map until the message has been disposed.

Comment (serbrech) [reply receiver.go:62]: I know it's not straight forward but are you planning to build the autocredit management on top of the manual credit, instead of aside of it?

File: internal/frames/parsing.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
+package frames
```
Comment (serbrech) [inline internal/frames/parsing.go:1]: This change seems unrelated. The PR is big enough as it is?

File: receiver.go [hunk 4]
```diff
@@ -108,39 +52,113 @@ func (r *Receiver) Receive(ctx context.Context) (*Message, error) {
...
 	case msg := <-r.link.Messages:
-		// we remove the message from unsettled map as soon as it's popped off the channel
-		// This makes the unsettled count the same as messages buffer count
-		// and keeps the behavior the same as before the unsettled messages tracking was introduced
-		defer r.link.DeleteUnsettled(&msg)
 		debug(3, "Receive() non blocking %d", msg.deliveryID)
-		msg.receiver = r
-		return &msg, nil
+		msg.link = r.link
+		return autoAccept(ctx, &msg)
```
Comment (serbrech) [inline receiver.go:72]: This gives the impression that we always autoaccept. I had to look at the implementation of autoaccept to see that it only accept if not mode second. Can probably use a better func name.

Comment (jhendrixMSFT) [reply receiver.go]: Good point, renamed.

File: internal/encoding/types.go [hunk 5]
```diff
@@ -343,7 +343,10 @@ func (s *SASLCode) Unmarshal(r *buffer.Buffer) error {
 	return err
 }
 
-type DeliveryState interface{} // TODO: http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transactions-v1.0-os.html#type-declared
+// TODO: http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transactions-v1.0-os.html#type-declared
+type DeliveryState interface {
+	deliveryState() // marker method
```
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply internal/encoding/types.go:348]: It provides some type-safety when dealing with polymorphic delivery states.  We use the same pattern for [frames](https://github.com/Azure/go-amqp/blob/master/internal/frames/frames.go#L342).

File: receiver.go (hunk 1 above)
Comment (jhendrixMSFT) [reply receiver.go]: Opened https://github.com/Azure/go-amqp/issues/73 to track consolidation.

File: receiver.go (hunk 2 above)
Comment (jhendrixMSFT) [reply receiver.go:42]: Added a check to not auto-accept the message when manual credit management is active.

File: integration_test.go [hunk 6]
```diff
@@ -381,21 +392,22 @@ func TestIntegrationReceiverModeSecond(t *testing.T) {
...
+							break
+						}
+						ctx, cancel = context.WithTimeout(context.Background(), 10*time.Second)
+						err = receiver.AcceptMessage(ctx, msg)
+						cancel()
+						if err != nil {
+							receiveErr.write(fmt.Errorf("Error accepting message: %+v", err))
+							break
+						}
+						if !bytes.Equal([]byte(data), msg.GetData()) {
```
Comment (richardpark-msft) [inline integration_test.go:408]: Similar thing here, this check could be right under Receive().

File: integration_test.go [hunk 7]
```diff
@@ -649,11 +661,14 @@ func TestIntegrationClose(t *testing.T) {
 
 		testClose(t, session.Close)
 
-		err = receiver.HandleMessage(context.Background(), nil)
+		msg, err := receiver.Receive(context.Background())
 		if err != amqp.ErrSessionClosed {
```
Comment (richardpark-msft) [inline integration_test.go:665]: You didn't change this but would the correct thing be to errors.Is() here as well? Or would convention say "this is never wrapped so you can compare it directly"?

File: internal/frames/parsing.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+	case encoding.TypeCodeSASLChallenge:
+		t := new(SASLChallenge)
+		err := t.Unmarshal(r)
+		return t, err
+	case encoding.TypeCodeSASLOutcome:
+		t := new(SASLOutcome)
+		err := t.Unmarshal(r)
+		return t, err
+	default:
+		return nil, fmt.Errorf("unknown preformative type %02x", pType)
```
Comment (richardpark-msft) [inline internal/frames/parsing.go:122]: NIT, but it's `performative`

File: internal/mocks/connection.go [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+type MockConnection struct {
+	// FrameCallback is invoked when an AMQP frame is received.
+	// Return a non-nil error to simulate a write error.
+	FrameCallback func(frames.FrameBody) error
+	readDLDur     time.Duration
+	readDL        *time.Timer
+	readData      chan []byte
+	pendingDataMu sync.Mutex
+	pendingData   [][]byte
+	//writeDL *time.Timer
```
Comment (richardpark-msft) [inline internal/mocks/connection.go:33]: NIT: remove comment.

But I am curious what you did with it....how cool is this MockConnection going to be? ;)

File: internal/mocks/connection.go [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+			Durable:      encoding.DurabilityNone,
+			ExpiryPolicy: encoding.ExpirySessionEnd,
+		},
+		ReceiverSettleMode: &mode,
+		MaxMessageSize:     math.MaxUint32,
+	})
+}
+
+// AppendTransfer appends a PerformTransfer frame with the specified values.
+// The linkHandle MUST match the linkHandle value specified in AppendReceiverAttach.
```
Comment (richardpark-msft) [inline internal/mocks/connection.go:107]: Perhaps at some point we'll bake that check into the MockConnection itself?

File: internal/mocks/connection.go [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+// following functions are for the net.Conn interface
+///////////////////////////////////////////////////////
+
+func (m *MockConnection) Read(b []byte) (n int, err error) {
+	if err := m.closedCheck(); err != nil {
+		return 0, err
+	}
+	select {
+	case <-m.readDL.C:
+		return 0, errors.New("read deadline exceeded")
```
Comment (richardpark-msft) [inline internal/mocks/connection.go:179]: NIT: Would it be possible to return the actual error you'd expect to see here? `net.OpError` perhaps?

Or are all errors (Temporary or Timeout) all treated the same? @serbrech had some pretty nice error handling code in go-shuttle that's made me think about this more.

File: internal/mocks/connection.go [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+	m.readData <- out
+	// swallow the write
+	return len(b), nil
+}
+
+func (m *MockConnection) Close() error {
+	m.closedMu.Lock()
+	defer m.closedMu.Unlock()
+	m.closed = true
+	if !m.readDL.Stop() {
```
Comment (richardpark-msft) [inline internal/mocks/connection.go:220]: I feel like it'd be useful to have a check that says "you shouldn't have anything left in the .readDL channel".

This might not be true in real world usage but in our tests I'm assuming it'd be a bug for us to leave behind any data.

File: internal/mocks/connection.go [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+
+func (m *MockConnection) LocalAddr() net.Addr {
+	return &net.IPAddr{
+		IP: net.IPv4(127, 0, 0, 1),
+	}
+}
+
+func (m *MockConnection) RemoteAddr() net.Addr {
+	return &net.IPAddr{
+		IP: net.IPv4(127, 0, 0, 1),
```
Comment (richardpark-msft) [inline internal/mocks/connection.go:234]: Just a question - would it be useful to mark the remote with a separate (but still loopbacky) address? Like the remote could be `127.0.0.2` just to distinguish it in our tests?

File: link.go [hunk 14]
```diff
@@ -95,7 +95,7 @@ func newLink(s *Session, r *Receiver, opts []LinkOption) (*link, error) {
 }
 
 // attachLink is used by Receiver and Sender to create new links
-func attachLink(s *Session, r *Receiver, opts []LinkOption) (*link, error) {
+func attachLink(ctx context.Context, s *Session, r *Receiver, opts []LinkOption) (*link, error) {
```
Comment (richardpark-msft) [inline link.go:98]: It's good to add a context here. Do you know if we have to unroll anything on our way out if we cancel though?

File: link.go [hunk 15]
```diff
@@ -227,7 +235,11 @@ func attachLink(s *Session, r *Receiver, opts []LinkOption) (*link, error) {
 		l.Messages = make(chan Message, l.receiver.maxCredit)
 		l.unsettledMessages = map[string]struct{}{}
 		// copy the received filter values
-		l.Source.Filter = resp.Source.Filter
+		if l.Source == nil {
```
Comment (richardpark-msft) [inline link.go:238]: Completeness or was this just an outright bug?

File: manualCreditor.go [hunk 16]
```diff
@@ -63,6 +62,10 @@ func (mc *manualCreditor) Drain(ctx context.Context) error {
 	// send drain, wait for responding flow frame
 	select {
 	case <-mc.drained:
+		// reset state here after the channel has been closed.
```
Comment (richardpark-msft) [inline manualCreditor.go:65]: I'm still not getting this. Is the problem that mc.drained might be nil before we reach line 65 because EndDrain happened so quickly?

Would it work the same if we just assigned the value of 'mc.drained' to a local 'drained' variable and then just <-'d on that? This would at least still make sure modifications to internals are within lock'd sections.

So, line 59ish:

```go
mc.drained = make(chan struct{})
drained := mc.drained
```

and then line 65ish:

```go
case <- drained:
```

File: receiver.go [hunk 17]
```diff
@@ -12,7 +12,7 @@ import (
 
 type messageDisposition struct {
 	id    uint32
-	state interface{}
+	state encoding.DeliveryState
```
Comment (richardpark-msft) [inline receiver.go:15]: Oh, ++ nice!

File: receiver.go [hunk 18]
```diff
@@ -108,39 +52,113 @@ func (r *Receiver) Receive(ctx context.Context) (*Message, error) {
 		}
 	}
 
+	// for ModeFirst, auto-accept the message
+	modeFirstAccept := func(ctx context.Context, msg *Message) (*Message, error) {
+		if r.manualCreditor != nil || receiverSettleModeValue(r.link.ReceiverSettleMode) == ModeSecond {
```
Comment (richardpark-msft) [inline receiver.go:57]: I don't think we should bypass modeFirstAccept for the manual creditor if I'm getting it (it's the equivalent of 'receive and delete', right?). 

If so, auto-accept is still correct in that mode, manual or not.

File: receiver.go [hunk 19]
```diff
@@ -108,39 +52,113 @@ func (r *Receiver) Receive(ctx context.Context) (*Message, error) {
...
+// Accept notifies the server that the message has been
+// accepted and does not require redelivery.
+func (r *Receiver) AcceptMessage(ctx context.Context, msg *Message) error {
+	if !msg.shouldSendDisposition() {
+		return nil
+	}
+	if err := r.messageDisposition(ctx, msg, &encoding.StateAccepted{}); err != nil {
+		return err
+	}
+	msg.settled = true
```
Comment (richardpark-msft) [inline receiver.go:103]: This would be useful as a public field, if you're willing. Or maybe a public function.

File: receiver.go [hunk 20]
```diff
@@ -108,39 +52,113 @@ func (r *Receiver) Receive(ctx context.Context) (*Message, error) {
...
+	return nil
+}
+
+// Reject notifies the server that the message is invalid.
+//
+// Rejection error is optional.
+func (r *Receiver) RejectMessage(ctx context.Context, msg *Message, e *Error) error {
```
Comment (richardpark-msft) [inline receiver.go:110]: Is there a reason it has to take an 'error'? Shouldn't it just be a string?

File: receiver.go [hunk 21]
```diff
@@ -108,39 +52,113 @@ func (r *Receiver) Receive(ctx context.Context) (*Message, error) {
...
+// deliveryFailed indicates that the server must consider this and
+// unsuccessful delivery attempt and increment the delivery count.
+//
+// undeliverableHere indicates that the server must not redeliver
+// the message to this link.
+//
+// messageAnnotations is an optional annotation map to be merged
+// with the existing message annotations, overwriting existing keys
+// if necessary.
+func (r *Receiver) ModifyMessage(ctx context.Context, msg *Message, deliveryFailed, undeliverableHere bool, messageAnnotations encoding.Annotations) error {
```
Comment (richardpark-msft) [inline receiver.go:146]: Do I have access to the `encoding.Annotations` type outside of go-amqp even though it's `internal`?

File: receiver_test.go [hunk 22]
```diff
@@ -2,137 +2,156 @@ package amqp
...
-		t.Errorf("HandleMessage() error = %v", err)
+	// link credit should be 1
+	if c := r.link.linkCredit; c != 1 {
+		t.Fatalf("unexpected link credit %d", c)
 	}
-
-	if len(r.dispositions) != 0 {
-		t.Errorf("the message should not have triggered a disposition")
+	// subsequent dispositions should have no effect
+	if err := r.AcceptMessage(context.Background(), msg); err != nil {
```
Comment (richardpark-msft) [inline receiver_test.go:68]: It seems like this should actually be an error. I can see no-oping at an application stack above this but it's a constraint violation to attempt to settle a message in mode first (or to settle a message twice, for instance).

File: receiver_test.go [hunk 23]
```diff
@@ -2,137 +2,156 @@ package amqp
...
+	if c := r.link.countUnsettled(); c != 1 {
+		t.Fatalf("unexpected unsettled count %d", c)
+	}
+	// wait for the link to pause as we've consumed all available credit
+	ctx, cancel = context.WithTimeout(context.Background(), 1*time.Second)
+	for {
+		if atomic.LoadUint32(&r.link.Paused) == 1 {
+			cancel()
+			break
+		} else if err := ctx.Err(); err != nil {
```
Comment (richardpark-msft) [inline receiver_test.go:116]: Do we have to do `errors.Is`?

File: receiver_test.go [hunk 24]
```diff
@@ -2,137 +2,156 @@ package amqp
...
+	msg, err := r.Receive(ctx)
+	cancel()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if c := r.link.countUnsettled(); c != 1 {
+		t.Fatalf("unexpected unsettled count %d", c)
+	}
+	// wait for the link to pause as we've consumed all available credit
+	ctx, cancel = context.WithTimeout(context.Background(), 1*time.Second)
```
Comment (richardpark-msft) [inline receiver_test.go:111]: btw, this is one of those instances I'm thinking something like https://github.com/Azure/go-amqp/issues/71 would be useful (do only a non-blocking select off the received messages channel).

File: receiver_test.go [hunk 25]
```diff
@@ -2,137 +2,156 @@ package amqp
...
-	// call handle with the accept handler in a goroutine because it will block on inflight disposition.
-	go func() {
-		if err := r.HandleMessage(context.TODO(), accept); err != nil {
-			t.Errorf("HandleMessage() error = %v", err)
+	// link credit must be zero since we only started with 1
+	if c := r.link.linkCredit; c != 0 {
+		t.Fatalf("unexpected link credit %d", c)
+	}
+	// create accepting disposition frame
+	assert.NoError(t, mock.AppendDisposition(deliveryID, &encoding.StateAccepted{}))
```
Comment (richardpark-msft) [inline receiver_test.go:125]: I really like what you've done here.

**Review by richardpark-msft — CHANGES_REQUESTED** [review]
There's some cosmetic things but had a few concerns.

File: example_test.go [hunk 26]
```diff
@@ -20,7 +20,7 @@ func Example() {
 	defer client.Close()
 
 	// Open a session
-	session, err := client.NewSession()
+	session, err := client.NewSession(context.Background())
```
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply example_test.go:23]: Good question, I hadn't thought about it much.  For examples, which customers tend to cut/paste, I think you're right.  That way, they should know to revisit it.

File: integration_test.go (hunk 7 above)
Comment (jhendrixMSFT) [reply integration_test.go:665]: A lot of this code predates `errors.Is()`, we should be using it.

File: receiver_test.go (hunk 22 above)
Comment (jhendrixMSFT) [reply receiver_test.go:68]: I couldn't find anything in the spec about this (possible I missed it).  Do you know what Rhea does here?  I can check .NET.

File: receiver_test.go (hunk 23 above)
Comment (jhendrixMSFT) [reply receiver_test.go:116]: No, just if it's non-nil means the context expired so we timed out.

File: internal/mocks/connection.go (hunk 11 above)
Comment (jhendrixMSFT) [reply internal/mocks/connection.go:179]: At present, all errors are treated as terminal, see [here](https://github.com/Azure/go-amqp/blob/master/conn.go#L530-L547).  Might be a candidate for some retry, but that can be handled separately.

Generally speaking, trying to reason about network errors beyond nil/non-nil falls apart pretty fast.  You initially start with three buckets of error types.

- errors we know to be transient - retry
- errors we know aren't transient - don't retry
- errors we are unsure about - retry

This seems reasonable on the surface, but in practice, if you don't own the chain of calls end-to-end, you can't distinguish case 1 and case 2, so everything coalesces into case 3, so you just always end up retrying.  This is why we added the NonRetriableError interface and annotate errors we _know_ are non-transient.  The vast majority are considered non-transient because you just don't know.

File: internal/mocks/connection.go (hunk 12 above)
Comment (jhendrixMSFT) [reply internal/mocks/connection.go:220]: Great idea!

File: link.go (hunk 14 above)
Comment (jhendrixMSFT) [reply link.go:98]: I don't believe so.  I say this because there were already other error cases that would return early and the link mux isn't started until the very end.

File: link.go (hunk 15 above)
Comment (jhendrixMSFT) [reply link.go:238]: It's a bug.  A handful of link options initialize `l.Source` and presumably we always use one or more of these in "real" code.  The receiver test code doesn't use any of these options which is how I hit this.

File: internal/mocks/connection.go (hunk 13 above)
Comment (jhendrixMSFT) [reply internal/mocks/connection.go:234]: Fixed

File: receiver.go (hunk 18 above)
Comment (jhendrixMSFT) [reply receiver.go:57]: I see, I misunderstood the comments in the other thread.  Will remove.

File: internal/mocks/connection.go (hunk 11 above)
Comment (serbrech) [reply internal/mocks/connection.go:179]: In goshuttle i progressively add error handling that identify when the connection inside the sdk is dead, and i reconstruct the publisher/listener in place. If I don't know the error yet, I let the service crash and restart. (A more brutal recover...)
That just forces me to be aware of the errors that do bubble out and I build up a robust in-process recovery layer over time.

File: receiver.go (hunk 21 above)
Comment (jhendrixMSFT) [reply receiver.go:146]: We export Annotations via a type alias, I fixed this to use that instead (good catch).

File: receiver.go (hunk 18 above)
Comment (serbrech) [reply receiver.go:57]: Correct mode first is receive and delete. You ack the receive before processing it.

File: receiver.go (hunk 20 above)
Comment (serbrech) [reply receiver.go:110]: +1 for using a string "reason" or something like this.

File: receiver_test.go (hunk 22 above)
Comment (serbrech) [reply receiver_test.go:68]: I agree. At the amqp level, I would make this an error. If you do settle twice, you'd get an error from the server too.
Would prefer a typed error though, so that we can check on on it safely.

File: receiver.go (hunk 20 above)
Comment (jhendrixMSFT) [reply receiver.go:110]: This is how the method originally was and conforms to the [spec](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#type-rejected).  If we do this, then there is some info callers can't specify.

The .NET [impl](https://github.com/Azure/azure-amqp/blob/59894e8d992c63dfcaad6d1d3f454d9d88855d91/src/ReceivingAmqpLink.cs#L363) follows the same pattern, so I'm inclined to leave it.

All that said, what about adding a convenience func to create the appropriate `Error` from a reason string or the like?

File: integration_test.go (hunk 7 above)
Comment (serbrech) [reply integration_test.go:665]: Errors.Is is go 1.13
go-amqp is still go 1.12

File: receiver.go (hunk 20 above)
Comment (serbrech) [reply receiver.go:110]: Leave it as is in this PR.

**Comment by serbrech** [comment]
The scope of the PR makes me a bit nervous. I would really like to see cleaner separation between refactoring PR and behavior change PRs

**Comment by jhendrixMSFT** [comment]
That's a fair point, let me look at splitting it up.

File: internal/mocks/connection.go (hunk 11 above)
Comment (richardpark-msft) [reply internal/mocks/connection.go]: @serbrech , track 2 service bus borrows most of that code so similar constraints/benefits there.

File: receiver_test.go (hunk 22 above)
Comment (jhendrixMSFT) [reply receiver_test.go]: This doesn't appear to be the case, i.e. I can accept a message multiple times without error.

File: receiver.go (hunk 19 above)
Comment (jhendrixMSFT) [reply receiver.go:103]: In hindsight I shouldn't have done this as this really isn't part of Message public surface area.  It was a simple way to prevent disposing a message multiple times which doesn't appear to be necessary.

Comment (jhendrixMSFT) [reply receiver.go:103]: If it does become problematic we can address it separately.

File: receiver.go [hunk 27]
```diff
@@ -273,17 +275,18 @@ func (r *Receiver) sendDisposition(first uint32, last *uint32, state encoding.De
...
 
-func (r *Receiver) messageDisposition(ctx context.Context, id uint32, state encoding.DeliveryState) error {
+func (r *Receiver) messageDisposition(ctx context.Context, msg *Message, state encoding.DeliveryState) error {
 	var wait chan error
 	if r.link.ReceiverSettleMode != nil && *r.link.ReceiverSettleMode == ModeSecond {
-		debug(3, "RX: add %d to inflight", id)
-		wait = r.inFlight.add(id)
+		debug(3, "RX: add %d to inflight", msg.deliveryID)
+		wait = r.inFlight.add(msg.deliveryID)
+		defer func() { r.link.DeleteUnsettled(msg) }()
```
Comment (jhendrixMSFT) [inline receiver.go:283]: Here's where we remove the message from the unsettled map.

File: receiver_test.go (hunk 22 above)
Comment (jhendrixMSFT) [reply receiver_test.go]: I somewhat misspoke.  While this works in ModeFirst, in ModeSecond, the second call to AcceptMessage() hangs :(

File: receiver.go (hunk 27 above)
Comment (richardpark-msft) [reply receiver.go:283]: Does the removal of the unsettled message depend on whether 'sendDisposition' below succeeds or fails? Was wondering why it was deferred (ie, which operation are you trying to get it to happen after?)

**Review by richardpark-msft — APPROVED** [review]
I think it looks good. 

The 'defer' part is interesting so it might just need a comment or something to explain what ordering you're hoping for (ie, this should happen after blah). 

I'm also curious if the operation failing impacts if we should remove the message from the unsettled map or not.

File: receiver.go (hunk 27 above)
Comment (jhendrixMSFT) [reply receiver.go:283]: I'm going over the spec and I think this might be wrong.  In section 2.6.12, I believe that we shouldn't be removing the entry from the unsettled map until we have final receipt of settlement from the sender.  i.e. we (the receiver) receive the message, send a disposition frame indicating we accept the message, the service sends a disposition frame acknowledging our acceptance, then we remove it from the unsettled map.  Do you concur?

Comment (richardpark-msft) [reply receiver.go:283]: I we're both reading that section the same way. 

1. We send a disposition
2. Sender acks it and sends _us_ a disposition 
3. Once we receive the ACK of our ACK (oof) then we can delete it from our map
